executor/join.go

```diff
@@ -296,7 +299,47 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
 			break
 		}
 
-		for _, outerRow := range outerBuffer.rows {
+		numMatchedOuters, numOuters := 0, len(outerBuffer.rows)
+		if e.outerFilter != nil {
+			outerFilterResult = outerFilterResult[:0]
+			for _, outerRow := range outerBuffer.rows {
+				matched, err := expression.EvalBool(e.outerFilter, outerRow, e.ctx)
+				if err != nil {
+					resultBuffer.err = errors.Trace(err)
+					break
+				}
+				outerFilterResult = append(outerFilterResult, matched)
+			}
+
+			for i, j := 0, len(outerFilterResult)-1; i < j; {
+				for i < j && outerFilterResult[i] == true {
+					i++
+				}
+				for i < j && outerFilterResult[j] == false {
+					j--
+				}
+				if i < j && outerFilterResult[i] == false && outerFilterResult[j] == true {
+					outerFilterResult[i], outerFilterResult[j] = outerFilterResult[j], outerFilterResult[i]
+					outerBuffer.rows[i], outerBuffer.rows[j] = outerBuffer.rows[j], outerBuffer.rows[i]
+					i++
+					j--
+				}
+			}
+			for numMatchedOuters < numOuters && outerFilterResult[numMatchedOuters] == true {
+				numMatchedOuters++
+			}
+		} else {
+			numMatchedOuters = numOuters
+		}
+
+		if len(outerBuffer.rows)-numMatchedOuters > 0 {
```

Comment:
I think we do not need this check,
if len(outerBuffer.rows)-numMatchedOuters == 0, `outerBuffer.rows[numMatchedOuters:]` will be `[]`